File: common/src/main/java/juuxel/adorn/compat/rei/BrewerDisplay.java
package juuxel.adorn.compat.rei;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import dev.architectury.fluid.FluidStack;
import juuxel.adorn.fluid.FluidIngredient;
import juuxel.adorn.fluid.FluidUnit;
import juuxel.adorn.platform.FluidBridge;
import juuxel.adorn.recipe.FluidBrewingRecipe;
import juuxel.adorn.recipe.ItemBrewingRecipe;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.display.DisplaySerializer;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.entry.EntryStack;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.Identifier;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Optional;

public record BrewerDisplay(
    EntryIngredient input,
    EntryIngredient first,
    EntryIngredient second,
    EntryIngredient fluid,
    EntryStack<?> result,
    Optional<Identifier> id
) implements Display {
    public static final MapCodec<BrewerDisplay> MAP_CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
        EntryIngredient.codec().fieldOf("input").forGetter(BrewerDisplay::input),
        EntryIngredient.codec().fieldOf("first_ingredient").forGetter(BrewerDisplay::first),
        EntryIngredient.codec().fieldOf("second_ingredient").forGetter(BrewerDisplay::second),
        EntryIngredient.codec().fieldOf("fluid").forGetter(BrewerDisplay::fluid),
        EntryStack.codec().fieldOf("result").forGetter(BrewerDisplay::result),
        Identifier.CODEC.optionalFieldOf("id").forGetter(BrewerDisplay::id)
    ).apply(instance, BrewerDisplay::new));

    public static final StreamCodec<RegistryFriendlyByteBuf, BrewerDisplay> PACKET_CODEC = StreamCodec.composite(
        EntryIngredient.streamCodec(), BrewerDisplay::input,
        EntryIngredient.streamCodec(), BrewerDisplay::first,
        EntryIngredient.streamCodec(), BrewerDisplay::second,
        EntryIngredient.streamCodec(), BrewerDisplay::fluid,
        EntryStack.streamCodec(), BrewerDisplay::result,
        ByteBufCodecs.optional(Identifier.STREAM_CODEC), BrewerDisplay::id,
        BrewerDisplay::new
    );

    public static final DisplaySerializer<BrewerDisplay> SERIALIZER = DisplaySerializer.of(MAP_CODEC, PACKET_CODEC);

    public BrewerDisplay(ItemBrewingRecipe recipe, @Nullable Identifier id) {
        this(
            EntryIngredients.ofIngredient(recipe.input()),
            EntryIngredients.ofIngredient(recipe.firstIngredient()),
            recipe.secondIngredient().map(EntryIngredients::ofIngredient).orElse(EntryIngredient.empty()),
            EntryIngredient.empty(),
            EntryStacks.of(recipe.result().create()),
            Optional.ofNullable(id)
        );
    }

    public BrewerDisplay(FluidBrewingRecipe recipe, @Nullable Identifier id) {
        this(
            EntryIngredients.ofIngredient(recipe.input()),
            EntryIngredients.ofIngredient(recipe.firstIngredient()),
            recipe.secondIngredient().map(EntryIngredients::ofIngredient).orElse(EntryIngredient.empty()),
            entryIngredientOf(recipe.fluid()),
            EntryStacks.of(recipe.result().create()),
            Optional.ofNullable(id)
        );
    }

    private static EntryIngredient entryIngredientOf(FluidIngredient fluidIngredient) {
        var amount = FluidUnit.convert(fluidIngredient.getAmount(), fluidIngredient.getUnit(), FluidBridge.get().getFluidUnit());
        var stacks = fluidIngredient.fluid()
            .getFluids()
            .stream()
            .map(fluid -> EntryStacks.of(FluidStack.create(fluid, amount, fluidIngredient.components())))
            .toList();
        return EntryIngredient.of(stacks);
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return List.of(input, first, second, fluid);
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.of(EntryIngredient.of(result));
    }

    @Override
    public CategoryIdentifier<?> getCategoryIdentifier() {
        return AdornReiServer.BREWER;
    }

    @Override
    public Optional<Identifier> getDisplayLocation() {
        return id;
    }

    @Override
    public DisplaySerializer<? extends Display> getSerializer() {
        return SERIALIZER;
    }
}

File: common/src/main/kotlin/juuxel/adorn/compat/rei/BrewerDisplay.kt
package juuxel.adorn.compat.rei

import juuxel.adorn.fluid.FluidUnit
import juuxel.adorn.item.AdornItems
import juuxel.adorn.platform.FluidBridge
import juuxel.adorn.recipe.ItemBrewingRecipe
import juuxel.adorn.recipe.FluidBrewingRecipe
import me.shedaniel.rei.api.common.category.CategoryIdentifier
import me.shedaniel.rei.api.common.display.Display
import me.shedaniel.rei.api.common.display.DisplaySerializer
import me.shedaniel.rei.api.common.entry.EntryIngredient
import me.shedaniel.rei.api.common.entry.EntryStack
import me.shedaniel.rei.api.common.util.EntryIngredients
import me.shedaniel.rei.api.common.util.EntryStacks
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement

class BrewerDisplay(
    val input: EntryIngredient,
    val first: EntryIngredient,
    val second: EntryIngredient,
    val fluid: EntryIngredient,
    val result: EntryStack<*>
) : Display {
    constructor(recipe: ItemBrewingRecipe) : this(
        EntryIngredients.of(AdornItems.MUG),
        first = EntryIngredients.ofIngredient(recipe.firstIngredient),
        second = EntryIngredients.ofIngredient(recipe.secondIngredient),
        fluid = EntryIngredient.empty(),
        result = EntryStacks.of(recipe.result)
    )

    constructor(recipe: FluidBrewingRecipe) : this(
        EntryIngredients.of(AdornItems.MUG),
        first = EntryIngredients.ofIngredient(recipe.firstIngredient),
        second = EntryIngredient.empty(),
        fluid = EntryIngredients.of(recipe.fluid.fluid, FluidUnit.convert(recipe.fluid.amount, recipe.fluid.unit, FluidBridge.get().fluidUnit)),
        result = EntryStacks.of(recipe.result)
    )

    override fun getInputEntries(): List<EntryIngredient> =
        listOf(input, first, second, fluid)

    override fun getOutputEntries(): List<EntryIngredient> =
        listOf(EntryIngredient.of(result))

    override fun getCategoryIdentifier(): CategoryIdentifier<*> =
        AdornReiServer.BREWER

    object Serializer : DisplaySerializer<BrewerDisplay> {
        override fun save(tag: NbtCompound, display: BrewerDisplay): NbtCompound = tag.apply {
            put("Input", display.input.save())
            put("FirstIngredient", display.first.save())
            put("SecondIngredient", display.second.save())
            put("Fluid", display.fluid.save())
            put("Result", display.result.save())
        }

        override fun read(tag: NbtCompound): BrewerDisplay {
            val input = EntryIngredient.read(tag.getList("Input", NbtElement.COMPOUND_TYPE.toInt()))
            val first = EntryIngredient.read(tag.getList("FirstIngredient", NbtElement.COMPOUND_TYPE.toInt()))
            val second = EntryIngredient.read(tag.getList("SecondIngredient", NbtElement.COMPOUND_TYPE.toInt()))
            val fluid = EntryIngredient.read(tag.getList("Fluid", NbtElement.COMPOUND_TYPE.toInt()))
            val result = EntryStack.read(tag.getCompound("Result"))
            return BrewerDisplay(input, first, second, fluid, result)
        }
    }
}
